src/Generator/Passes/DelegatesPass.cs

```diff
@@ -89,39 +105,68 @@ public override bool VisitMethodDecl(Method method)
             }
 
             var @delegate = new TypedefDecl
-                {
-                    Name = delegateName,
-                    QualifiedType = new QualifiedType(
+            {
+                Name = delegateName,
+                QualifiedType = new QualifiedType(
                         new PointerType(
                             new QualifiedType(
                                 new FunctionType
                                 {
-                                    CallingConvention = method.CallingConvention,
-                                    IsDependent = method.IsDependent,
+                                    CallingConvention = callingConvention,
+                                    IsDependent = isDependent,
                                     Parameters = @params,
-                                    ReturnType = method.ReturnType
+                                    ReturnType = returnType
                                 }))),
-                    Namespace = namespaceDelegates,
-                    IsSynthetized = true,
-                    Access = AccessSpecifier.Private
-                };
+                Namespace = namespaceDelegates,
+                IsSynthetized = true,
+                Access = (decl is Method) ? AccessSpecifier.Private : AccessSpecifier.Public
+            };
 
             var delegateString = @delegate.Visit(TypePrinter).Type;
             var existingDelegate = GetExistingDelegate(
-                method.TranslationUnit.Module, delegateString);
+                module, delegateString);
 
             if (existingDelegate != null)
             {
-                Context.Delegates.Add(method, existingDelegate);
-                return true;
+                Context.Delegates.Add(decl, existingDelegate);
+                return;
             }
 
             existingDelegate = new DelegateDefinition(module.OutputNamespace, delegateString);
-            Context.Delegates.Add(method, existingDelegate);
+            Context.Delegates.Add(decl, existingDelegate);
 
             libsDelegates[module].Add(delegateString, existingDelegate);
 
             namespaceDelegates.Declarations.Add(@delegate);
+        }
+
+        private void VisitFunctionTypeParameters(Method method)
+        {
+            foreach(var param in method.Parameters)
+            {
+                if (param.QualifiedType.Type.GetFinalQualifiedPointee().Type.Desugar() is FunctionType)
```

Comment:
This check is both too long and incomplete. First you need to ensure you actually have a pointee by checking the type is an address:

    var paramType = param.Type.Desugar();
    paramType.IsAddress()

Then shorten the rest to:

    paramType.GetFinalPointee().Desugar() is FunctionType

Also, in this particular case change the condition to LINQ, use its SQL syntax.